src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,277 @@
+        $queryExtraKeys = array_diff(array_keys($query), array_keys($configuration));
+        if (0 < \count($queryExtraKeys)) {
+            throw new TransportException(sprintf('Unknown option found in DSN: [%s]. Allowed options are [%s]', implode(', ', $queryExtraKeys), implode(', ', self::DEFAULT_OPTIONS)));
+        }
+
+        return $configuration;
+    }
+
+    /**
+     * @param int $delay The delay in milliseconds
+     *
+     * @throws \Doctrine\DBAL\DBALException
+     */
+    public function send(string $body, array $headers, int $delay = 0): void
+    {
+        $now = (\DateTime::createFromFormat('U.u', microtime(true)));
+        $availableAt = (clone $now)->modify(sprintf('+%d seconds', $delay / 1000));
+
+        $values = [
+            'body' => ':body',
+            'headers' => ':headers',
+            'queue_name' => ':queue_name',
+            'created_at' => ':created_at',
+            'available_at' => ':available_at',
+        ];
+        if (self::ID_STRATEGY_UUID === $this->configuration['id_strategy']) {
+            $values['id'] = $this->driverConnection->getDatabasePlatform()->getGuidExpression();
+        }
+        $queryBuilder = $this->driverConnection->createQueryBuilder()
+            ->insert($this->configuration['table_name'])
+            ->values($values);
+
+        $this->executeQuery($queryBuilder->getSQL(), [
+            ':body' => $body,
+            ':headers' => Type::getType(Type::JSON)->convertToDatabaseValue($headers, $this->driverConnection->getDatabasePlatform()),
+            ':queue_name' => $this->configuration['queue_name'],
+            ':created_at' => $now->format('Y-m-d H:i:s.u'),
+            ':available_at' => $availableAt->format('Y-m-d H:i:s.u'),
+        ]);
+    }
+
+    public function get(): ?array
+    {
+        $this->driverConnection->beginTransaction();
+        try {
+            $query = $this->driverConnection->createQueryBuilder()
+                ->select('m.*')
+                ->from($this->configuration['table_name'], 'm')
+                ->where('m.delivered_at is null OR m.delivered_at < :redeliver_limit')
+                ->andWhere('m.available_at <= :now')
+                ->andWhere('m.queue_name = :queue_name')
+                ->orderBy('available_at', 'ASC')
+                ->setMaxResults(1);
+
+            $now = \DateTime::createFromFormat('U.u', microtime(true));
+            $redeliverLimit = (clone $now)->modify(sprintf('-%d seconds', $this->configuration['redeliver_timeout']));
+            // use SELECT ... FOR UPDATE to lock table
+            $doctrineEnvelope = $this->executeQuery(
+                $query->getSQL().' '.$this->driverConnection->getDatabasePlatform()->getWriteLockSQL(),
+                [
+                    ':now' => $now->format('Y-m-d H:i:s.u'),
+                    ':queue_name' => $this->configuration['queue_name'],
+                    ':redeliver_limit' => $redeliverLimit->format('Y-m-d H:i:s'),
```

Comment:
This highlights the need for the `formatDateTime` method so we are sure *all* the formats are the same.